### ISSUE TITLE
[Feature]: consider withTrashed when inferring Model properties on related SoftDeleted Models

### ISSUE THREAD

**Issue by ayrtonandino** [calebdw/phpstan-laravel#27]
# Feature Request

im using `_ide_helper_models.php` to define Model properties, as per documentation [phpstan-laravel Laravel IDE Helper](https://phpstan-laravel.dev/1.x/guide/model-properties/#laravel-ide-helper) i remove _ide_helper_models from scanFiles parameter, and got **456 errors**, the vast majority must be false positives coming from Relations using SoftDeletes.

For example, this phpstan error:

```txt
------ ------------------------------------------------------------------------------ 
  Line   app/Http/Controllers/CommissionsController.php                                
 ------ ------------------------------------------------------------------------------ 
  43     Cannot access property $award_calculation_level on App\Models\Vendedor|null.  
         🪪  property.nonObject                                                        
         ✏️  app/Http/Controllers/CommissionsController.php:43                         
 ------ ------------------------------------------------------------------------------
```

This comes from accessing the relation BelongsTo<Vendedor> ($model->vendedor), Vendedor uses SoftDeletes so phpstan thinks it could be null (App\Models\Vendedor|null).

soft-deleted records are excluded from queries by default, meaning even non-nullable foreign keys can return null when the related model is soft-deleted.

In this case, this relation will **never be null**, because:

Its key is not nullable and constrained: 

```php
    // migration
    $table->foreignIdFor(Vendedor::class)->constrained('vendedores')->cascadeOnUpdate();
```

And its relation is created withTrashed:

```php
    // model
    /**
     * @return BelongsTo<Vendedor, $this>
     */
    public function vendedor(): BelongsTo
    {
        return $this->belongsTo(Vendedor::class)->withTrashed();
    }
```

I don't know what definition is being generated right now, but in this case it should be:

```php
/**
     * @property int $id
     * @property int $vendedor_id
     * @property-read Vendedor $vendedor
*/
```

**Comment by calebdw** [comment]
Thanks for opening this issue and for including a concrete example. I understand why `withTrashed()` combined with a non-nullable, constrained foreign key appears to make this relationship non-nullable.

I’m going to close this as not planned for now, but I want to explain the reasoning and the current technical limitations rather than simply dismissing the idea.

Laravel’s base `Relation` class already has three template types:

```php
Relation<TRelatedModel, TDeclaringModel, TResult>
```

For `BelongsTo`, `TRelatedModel` represents the model used by the relationship and its query builder, while `TResult` represents the value returned when the relationship property is accessed. Those are deliberately separate concepts: the related model and builder cannot be nullable, but the result of resolving the relationship can be.

Conceptually, `BelongsTo` should expose `TResult` with a nullable default:

```php
/**
 * @template TRelatedModel of Model
 * @template TDeclaringModel of Model
 * @template TResult = TRelatedModel|null
 *
 * @extends Relation<TRelatedModel, TDeclaringModel, TResult>
 */
class BelongsTo extends Relation
{
}
```

The default is important because nullable is the conservative and overwhelmingly common behavior. Users should not have to repeat the third template argument on every ordinary `BelongsTo` declaration.

Unfortunately, PHPStan does not currently support referencing one class template from another template’s default value correctly. That limitation is tracked in [phpstan/phpstan#12674](https://github.com/phpstan/phpstan/issues/12674), where the expected behavior has been acknowledged as something PHPStan should eventually support.

If that limitation is resolved, it may become possible to represent operations such as `withDefault()` accurately:

```php
BelongsTo<Vendedor, $this, Vendedor|null>
```

could become:

```php
BelongsTo<Vendedor, $this, Vendedor>
```

That transformation would be sound because `withDefault()` constructs a new related model when no database record is found, so the relationship result is no longer nullable.

`withTrashed()` is different. It does not guarantee that a related model exists; it only removes the `SoftDeletingScope` from the related query. It is also not a method declared directly on `BelongsTo`. Laravel registers it dynamically as a builder macro through `SoftDeletingScope`, and `Relation::__call()` forwards the invocation to the underlying Eloquent builder. Consequently, there is no relationship method signature where a different `TResult` could simply be declared.

A specialized PHPStan extension might be able to recognize some direct `withTrashed()` calls, but changing the result to non-nullable would still not be generally sound. `withTrashed()` only removes one reason the query might return no result. For example:

```php
$model = new Commission();

$model->vendedor; // null
```

That remains true regardless of the database column definition because this model has not been persisted and has no foreign key value.

There are other possible runtime states as well:

- The related query can contain additional constraints.
- Another global scope can exclude the related record.
- The relation can be assigned or cleared in memory.
- The model can contain an invalid or stale foreign key before it is persisted.
- Database constraints describe persisted database state, not every possible state of an Eloquent model in PHP.

The extension also cannot determine from this relationship definition alone that the foreign key is both non-nullable and constrained. Doing that would require resolving the relationship’s effective foreign key, correlating it with the parsed database schema, interpreting the forwarded `withTrashed()` call, and then assuming that those database guarantees apply to the model’s current in-memory state. Even after all of that, the `new Model()` example would remain nullable.

For those reasons, the inferred nullable type is intentional. It is safer for PHPStan to require one redundant check when the application has a stronger invariant than to infer a non-null type that can still produce `null` at runtime.

When this is an application-level invariant, the recommended solution is to narrow the type explicitly:

```php
$vendedor = $model->vendedor;

assert($vendedor !== null);

$vendedor->award_calculation_level;
```

PHPStan understands the assertion and treats `$vendedor` as `Vendedor` afterward. This also documents and checks the invariant exactly where the code relies upon it.

If you want the invariant enforced even when PHP assertions are disabled, use an explicit exception instead:

```php
$vendedor = $model->vendedor;

if ($vendedor === null) {
    throw new LogicException('Commission must have a vendedor.');
}

$vendedor->award_calculation_level;
```

If a fallback model is appropriate for the application, `withDefault()` is another option:

```php
public function vendedor(): BelongsTo
{
    return $this->belongsTo(Vendedor::class)->withTrashed()->withDefault();
}
```

However, that changes runtime behavior and may hide a broken invariant, so it should only be used when a default model is genuinely desired.

This is closely related to [larastan/larastan#2335](https://github.com/larastan/larastan/issues/2335), which discusses why nullability belongs in `TResult` rather than `TRelatedModel`, and why narrowing is currently the safest approach.

Thanks again for raising this and providing the example. The use case is reasonable, but the combination of PHPStan’s current generic-template limitation, Laravel’s dynamically forwarded `withTrashed()` macro, and the fact that `withTrashed()` does not actually guarantee a result means this cannot be modeled safely at present.

**Comment by ayrtonandino** [comment]
Thanks for taking the time to explain, really interesting reading.

I'm going to add some throw_if since there are many more vectors that I never considered. Maybe create a new throw_if_null guard for simplicity.

**Comment by calebdw** [comment]
something like this would probably work, and let PHPStan know that the value cannot be nullable after the fact

```php
/**
 * @template T
 * @param T $value
 * @return (T is null ? never : void)
 */
function throw_if_null(mixed $value): void
{
    throw_if($value === null, 'Value cannot be null');
}
```

Or you can also use https://github.com/webmozarts/assert and do something like:

```php
Assert::notNull();
```
